operator/src/main/java/oracle/kubernetes/operator/helpers/DomainValidationSteps.java

```diff
@@ -173,103 +165,19 @@ static class ValidateDomainTopologyStep extends Step {
-        logAndAddValidationWarning(info, MessageKeys.NO_CLUSTER_IN_DOMAIN, clusterName);
-      }
-    }
-
-    private void warnIfServerDoesNotExist(WlsDomainConfig domainConfig,
-        String serverName, DomainPresenceInfo info) {
-      if (!domainConfig.containsServer(serverName)) {
-        logAndAddValidationWarning(info, MessageKeys.NO_MANAGED_SERVER_IN_DOMAIN, serverName);
-      }
-    }
-
-    private void warnIfMonitoringExporterPortConflicts(
-        Integer port, WlsServerConfig serverConfig, DomainPresenceInfo info) {
-      warnIfMonitoringExporterPortConflicts(port, null, serverConfig, info);
-    }
-
-    private void warnIfMonitoringExporterPortConflicts(
-        Integer port, WlsClusterConfig cluster, WlsServerConfig serverConfig, DomainPresenceInfo info) {
-
-      if (port.equals(serverConfig.getListenPort())) {
-        logAndAddValidationWarningExporter(port, cluster, serverConfig, serverConfig.getListenPort(), info);
-      }
-      if (port.equals(serverConfig.getSslListenPort())) {
-        logAndAddValidationWarningExporter(port, cluster, serverConfig, serverConfig.getSslListenPort(), info);
-      }
-      if (port.equals(serverConfig.getAdminPort())) {
-        logAndAddValidationWarningExporter(port, cluster, serverConfig, serverConfig.getAdminPort(), info);
-      }
-      Optional.ofNullable(serverConfig.getNetworkAccessPoints()).ifPresent(naps -> naps.forEach(nap -> {
-        if (port.equals(nap.getListenPort())) {
-          logAndAddValidationWarningExporter(port, cluster, serverConfig, nap.getListenPort(), info);
-        }
-      }));
-    }
-
-    private void logAndAddValidationWarningExporter(
-        Integer port, WlsClusterConfig cluster, WlsServerConfig serverConfig,
-        Integer conflictPort, DomainPresenceInfo info) {
-      if (cluster != null) {
-        logAndAddValidationWarning(info, MessageKeys.MONITORING_EXPORTER_CONFLICT_DYNAMIC_CLUSTER,
-            port, cluster.getClusterName(), conflictPort);
-      } else {
-        logAndAddValidationWarning(info, MessageKeys.MONITORING_EXPORTER_CONFLICT_SERVER,
-            port, serverConfig.getName(), conflictPort);
-      }
-    }
-
     @Override
     public NextAction apply(Packet packet) {
-      DomainPresenceInfo info = packet.getSpi(DomainPresenceInfo.class);
-      boolean isExplicitRecheck = MakeRightDomainOperation.isExplicitRecheck(packet);
-      WlsDomainConfig wlsDomainConfig = (WlsDomainConfig) packet.get(ProcessingConstants.DOMAIN_TOPOLOGY);
-      validate(info, wlsDomainConfig);
-
-      return doNext(getNextStep(info.getValidationWarningsAsString(), isExplicitRecheck, getNext()), packet);
+      List<String> failures = new WlsConfigValidator(packet).loggingTo(LOGGER).getFailures();
+      if (failures.isEmpty()) {
+        return doNext(createRemoveSelectedFailuresStep(getNext(), TOPOLOGY_MISMATCH), packet);
+      } else {
+        return doNext(createTopologyMismatchFailureSteps(getNext(), failures), packet);
```

Comment:
Is it possible that the domain already has a TopologyMistach failure condition from the previous make-right operation? If so, we may need to remove the old one before adding a new one.

Reply:
In the test, we update the topology and the domain and then compare the new domain with the old topology, which are expected to be incompatible. We're not supposed to be comparing them. It is possible that failing to call the correct FailureStep constructor is actually causing the test to fail, but the logic needs to be fixed in any case.

Reply:
My comment is general, not related to any test cases.  
If we have a TopologyMismatch failure in the previous make-right, what should we do?

Reply:
TopologyValidationStepTest has two tests that address this issue:

``removeOldToplogyFailures`` verifies that when the validation step is run, if there were topology failures from the previous make-right and they have since been corrected, they are removed.

``preserveTopologyFailuresThatStillExist`` verifies that if the problems have not been corrected, the old failures are not removed, and the original failure time is preserved.

This is the pattern that will be used for all failures, to avoid the flickering issue you found previously. Rather than clearing all failures at the start of the make-right, it is the responsibility of steps which check for failures to remove those that no longer exist. I'm not sure yet how to handle Kubernetes failures, but I think this can be made to work for ServerPod failures - and it certainly works for state-based failures such as DomainInvalid, TopologyMismatch, Introspection and ReplicasTooHIgh.

Reply:
I am not suggesting to remove all failures at the start of the make-right. I was thinking about a situation where the domain failed with some topology mismatch errors, and then the user changed the domain to fix the problem, but still has some errors remaining, or introduced new errors.  The remaining and new errors will be reported. The question is what we need to do with the original/previous failure condition. I think it is confusing to leave the stale failure condition in the domain status because errors in the original failure condition do not apply to the current version of the domain spec any more.

Reply:
Are you saying that the old Failed/TopologyMismatch condition is removed when the new Failed/TopologyMismatch condition is added? I don't see where it is done.

Reply:
Yes, look at DomainStatusUpdater.createTopologyMismatchSteps(). It appends a call to ``removingOldFailures(TOPOLOGY_MISMATCH)`` which is then used to select existing failures for removal.

Reply:
I see. thanks.